File: src/mcp/references/refGrammar.ts
import { lex, type Token, tokensToText } from "./refLexer.js";

////////////////////////////////
//  Interfaces & Types

/** W3C Text Fragment semantics, with one deviation: `@before`/`@after` mean nearest-in-scope, not
 * the spec's immediate adjacency. */
export type Matcher =
	| { kind: "text"; text: string }
	| { kind: "before"; text: string; anchor: string }
	| { kind: "after"; text: string; anchor: string }
	| { kind: "range"; from: string; to: string };

/** Every string here is DECODED; only `canonicalKey` re-encodes. */
export interface Ref {
	/** As written. Confinement is the resolver's job. */
	path: string;
	/** Each matches any descendant of the previous, not only a direct child. */
	segments: string[];
	matcher: Matcher | null;
}

export type ParseErrorCode =
	| "path-required"
	| "empty-fragment"
	| "empty-match-text"
	| "empty-anchor"
	| "empty-range-bound";

/** Parsing is TOTAL: a ref, not a ref, or a stated error at a stated offset. There is no outcome
 * where a malformed ref silently becomes a different valid one. */
export type ParseResult =
	| { kind: "ok"; ref: Ref }
	| { kind: "not-a-ref" }
	| { kind: "error"; code: ParseErrorCode; message: string; offset: number };

////////////////////////////////
//  Functions & Helpers

export const REF_SCHEME = "ref://";

const ANCHOR_KEYWORDS = ["before", "after"] as const;

type AnchorKeyword = (typeof ANCHOR_KEYWORDS)[number];

// What the reader treats specially: structural tokens, `%`, and what tryParseRef strips.
const SCOPE_ESCAPES = new Set(["%", ":", "#", "<", ">"]);
const FRAGMENT_ESCAPES = new Set(["%", ":", "#", ".", "@", "<", ">"]);

function escapeChar(c: string): string {
	return [...new TextEncoder().encode(c)].map((b) => `%${b.toString(16).toUpperCase().padStart(2, "0")}`).join("");
}

function encode(raw: string, escapes: Set<string>): string {
	let out = "";
	for (const c of raw) out += escapes.has(c) || /\s/.test(c) ? escapeChar(c) : c;
	return out;
}

function error(code: ParseErrorCode, message: string, offset: number): ParseResult {
	return { kind: "error", code, message, offset };
}

/** Only that exact shape is structural, so an `@` in an email address needs no encoding. */
function findAnchor(tokens: Token[]): { at: number; keyword: AnchorKeyword; after: number } | null {
	for (let i = 0; i < tokens.length; i++) {
		if (tokens[i].kind !== "at") continue;
		for (const keyword of ANCHOR_KEYWORDS) {
			const end = i + 1 + keyword.length;
			if (tokens[end]?.kind !== "sep") continue;
			const spelled = tokens
				.slice(i + 1, end)
				.map((t) => (t.kind === "char" ? t.value : ""))
				.join("");
			if (spelled === keyword) return { at: i, keyword, after: end + 1 };
		}
	}
	return null;
}

type MatcherResult = ParseResult | { kind: "matcher"; matcher: Matcher };

/** The FIRST structural marker by position wins and the rest is ordinary text, so a second `@after:`
 * needs no encoding and the form never depends on branch order. */
function parseMatcher(tokens: Token[], hashOffset: number): MatcherResult {
	if (tokens.length === 0) {
		return error("empty-fragment", `a \`#\` with nothing after it selects nothing`, hashOffset);
	}

	const anchor = findAnchor(tokens);
	const rangeAt = tokens.findIndex((t) => t.kind === "range");

	if (anchor && (rangeAt === -1 || anchor.at < rangeAt)) {
		const text = tokensToText(tokens.slice(0, anchor.at));
		const anchorText = tokensToText(tokens.slice(anchor.after));
		if (text === "") {
			return error("empty-match-text", `nothing to search for before the anchor`, tokens[0].offset);
		}
		if (anchorText === "") {
			return error(
				"empty-anchor",
				`\`@${anchor.keyword}:\` needs text to anchor against`,
				tokens[anchor.at].offset,
			);
		}
		return { kind: "matcher", matcher: { kind: anchor.keyword, text, anchor: anchorText } };
	}

	if (rangeAt !== -1) {
		const from = tokensToText(tokens.slice(0, rangeAt));
		const to = tokensToText(tokens.slice(rangeAt + 1));
		if (from === "" || to === "") {
			return error("empty-range-bound", `a range needs text on both sides of \`..\``, tokens[rangeAt].offset);
		}
		return { kind: "matcher", matcher: { kind: "range", from, to } };
	}

	return { kind: "matcher", matcher: { kind: "text", text: tokensToText(tokens) } };
}

/** Not `new URL()`: under URL rules `ref://app.js:Foo` reads `app.js` as a host and `:Foo` as a
 * port. */
export function tryParseRef(uri: string): ParseResult {
	// Markdown's wrapper comes off only as a PAIR, or `#Promise<Response>` silently shortens.
	const bare = uri.trim();
	const wrapped = bare.length >= 2 && bare.startsWith("<") && bare.endsWith(">");
	const unwrapped = wrapped ? bare.slice(1, -1) : bare;
	if (!unwrapped.toLowerCase().startsWith(REF_SCHEME)) return { kind: "not-a-ref" };

	// Shifted past the scheme, so an offset indexes the ref as WRITTEN.
	const tokens = lex(unwrapped.slice(REF_SCHEME.length)).map((t) => ({ ...t, offset: t.offset + REF_SCHEME.length }));
	const hashAt = tokens.findIndex((t) => t.kind === "hash");
	const scope = hashAt === -1 ? tokens : tokens.slice(0, hashAt);

	// path := char+, required, or `ref://:Foo` promotes its first segment into the path slot.
	const firstSep = scope.findIndex((t) => t.kind === "sep");
	const pathTokens = firstSep === -1 ? scope : scope.slice(0, firstSep);
	if (pathTokens.length === 0) {
		return error("path-required", `a ref needs a file path before any \`:\``, 0);
	}

	// segment := char*, so an empty one merges. That IS the `::` collapse.
	const segments: string[] = [];
	let cursor = firstSep;
	while (cursor !== -1) {
		const nextSep = scope.findIndex((t, i) => i > cursor && t.kind === "sep");
		const text = tokensToText(scope.slice(cursor + 1, nextSep === -1 ? scope.length : nextSep));
		if (text !== "") segments.push(text);
		cursor = nextSep;
	}

	const ref: Ref = { path: tokensToText(pathTokens), segments, matcher: null };
	if (hashAt === -1) return { kind: "ok", ref };

	const matcher = parseMatcher(tokens.slice(hashAt + 1), tokens[hashAt].offset);
	if (matcher.kind !== "matcher") return matcher;
	return { kind: "ok", ref: { ...ref, matcher: matcher.matcher } };
}

/** Null for anything malformed. Use `tryParseRef` when the caller needs the reason. */
export function parseRef(uri: string): Ref | null {
	const result = tryParseRef(uri);
	return result.kind === "ok" ? result.ref : null;
}

function serializeMatcher(matcher: Matcher): string {
	const text = (raw: string) => encode(raw, FRAGMENT_ESCAPES);
	switch (matcher.kind) {
		case "text":
			return `#${text(matcher.text)}`;
		case "before":
		case "after":
			return `#${text(matcher.text)}@${matcher.kind}:${text(matcher.anchor)}`;
		case "range":
			return `#${text(matcher.from)}..${text(matcher.to)}`;
	}
}

/** A ref's stable identity, which the console recomputes from a tapped link. Idempotent by
 * construction: everything the reader treats specially is escaped. */
export function canonicalKey(ref: Ref): string {
	const scope = [ref.path, ...ref.segments].map((c) => encode(c, SCOPE_ESCAPES)).join(":");
	return `${REF_SCHEME}${scope}${ref.matcher ? serializeMatcher(ref.matcher) : ""}`;
}

/** Parse and canonicalize in one step, or null if the input is not a well-formed ref. */
export function canonicalizeUri(uri: string): string | null {
	const ref = parseRef(uri);
	return ref ? canonicalKey(ref) : null;
}

File: android/app/src/main/java/com/atelier_nyaarium/switchboard/proto/RefGrammar.kt
package com.atelier_nyaarium.switchboard.proto

import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.CodingErrorAction

/**
 * Hand-authored twin of `src/mcp/references/refLexer.ts` + `refGrammar.ts`.
 *
 * The MCP writes canonical ref keys into a message's manifest; this side recomputes a key from a
 * tapped link and looks it up. If the two ever disagree on one character, every tap on that ref
 * misses and nothing reports why, so the twin is held to the TS source by the shared vector corpus
 * (`tests/fixtures/refs/vectors.json`, read by `RefGrammarVectorsTest` and by vitest).
 *
 * Two lexing modes, five token kinds, and one ambiguity rule: the first structural marker by
 * position wins and everything after it is ordinary text. The full spec lives in
 * `plans/artifact-references.md`.
 */
sealed class Matcher {
	data class Text(val text: String) : Matcher()

	data class Before(val text: String, val anchor: String) : Matcher()

	data class After(val text: String, val anchor: String) : Matcher()

	data class Range(val from: String, val to: String) : Matcher()
}

data class Ref(val path: String, val segments: List<String>, val matcher: Matcher?)

sealed class RefParseResult {
	data class Ok(val ref: Ref) : RefParseResult()

	object NotARef : RefParseResult()

	/** `code` matches the TS `ParseErrorCode` union exactly; the vectors pin both it and the offset. */
	data class Error(val code: String, val offset: Int) : RefParseResult()
}

const val REF_SCHEME = "ref://"

private enum class Kind { SEP, HASH, RANGE, AT, CHAR }

private data class Token(val kind: Kind, val value: String, val offset: Int)

private val ANCHOR_KEYWORDS = listOf("before", "after")

// Derived from what the reader treats specially in each mode, never hand-kept: that mode's
// structural characters, `%` because it introduces an escape, and `<`/`>` plus whitespace because
// the parser strips those before lexing.
private val SCOPE_ESCAPES = setOf('%', ':', '#', '<', '>')
private val FRAGMENT_ESCAPES = setOf('%', ':', '#', '.', '@', '<', '>')

private fun isHex(c: Char) = c in '0'..'9' || c in 'a'..'f' || c in 'A'..'F'

/**
 * Decode a run of `%XX` escapes as one unit, or null if the bytes are not valid UTF-8.
 *
 * REPORT rather than the platform default: a plain `String(bytes, UTF_8)` substitutes U+FFFD and
 * succeeds, where the TS side's `decodeURIComponent` throws and falls back to literal text. Matching
 * that requires asking the decoder to fail.
 */
private fun decodeEscapeRun(bytes: ByteArray): String? {
	val decoder = Charsets.UTF_8.newDecoder()
		.onMalformedInput(CodingErrorAction.REPORT)
		.onUnmappableCharacter(CodingErrorAction.REPORT)
	return try {
		decoder.decode(ByteBuffer.wrap(bytes)).toString()
	} catch (_: CharacterCodingException) {
		null
	}
}

/** Every code point of `s`, so an astral character is one unit rather than two surrogates. */
private fun codePoints(s: String): List<String> {
	val out = ArrayList<String>(s.length)
	var i = 0
	while (i < s.length) {
		val cp = s.codePointAt(i)
		val width = Character.charCount(cp)
		out.add(s.substring(i, i + width))
		i += width
	}
	return out
}

private fun lex(input: String): List<Token> {
	val tokens = ArrayList<Token>()
	var fragment = false
	var i = 0

	while (i < input.length) {
		val c = input[i]

		if (c == '%') {
			var end = i
			val bytes = ArrayList<Byte>()
			while (end + 2 < input.length && input[end] == '%' && isHex(input[end + 1]) && isHex(input[end + 2])) {
				bytes.add(input.substring(end + 1, end + 3).toInt(16).toByte())
				end += 3
			}
			val decoded = if (bytes.isEmpty()) null else decodeEscapeRun(bytes.toByteArray())
			if (decoded != null) {
				for (unit in codePoints(decoded)) tokens.add(Token(Kind.CHAR, unit, i))
				i = end
				continue
			}
			tokens.add(Token(Kind.CHAR, "%", i))
			i++
			continue
		}

		if (c == ':') {
			tokens.add(Token(Kind.SEP, "", i))
			i++
			continue
		}

		if (c == '#' && !fragment) {
			tokens.add(Token(Kind.HASH, "", i))
			fragment = true
			i++
			continue
		}

		if (fragment && c == '@') {
			tokens.add(Token(Kind.AT, "", i))
			i++
			continue
		}

		if (fragment && c == '.') {
			var run = i
			while (run < input.length && input[run] == '.') run++
			val length = run - i
			// Exactly two dots is a range. Three or more is a spread or a variadic, so `#...args`
			// searches for that text instead of parsing as a broken range.
			if (length == 2) {
				tokens.add(Token(Kind.RANGE, "", i))
			} else {
				for (n in 0 until length) tokens.add(Token(Kind.CHAR, ".", i + n))
			}
			i = run
			continue
		}

		// Advance by code point so an astral character stays one token, matching the JS iterator.
		val cp = input.codePointAt(i)
		val width = Character.charCount(cp)
		tokens.add(Token(Kind.CHAR, input.substring(i, i + width), i))
		i += width
	}

	return tokens
}

private fun tokenText(token: Token): String =
	when (token.kind) {
		Kind.SEP -> ":"
		Kind.HASH -> "#"
		Kind.RANGE -> ".."
		Kind.AT -> "@"
		Kind.CHAR -> token.value
	}

private fun tokensToText(tokens: List<Token>): String = tokens.joinToString("") { tokenText(it) }

private data class Anchor(val at: Int, val keyword: String, val after: Int)

/** Where an anchor marker starts, or null. Only `@` + exactly `before`/`after` + `:` is structural,
 * so an at-sign in an email address needs no encoding. */
private fun findAnchor(tokens: List<Token>): Anchor? {
	for (i in tokens.indices) {
		if (tokens[i].kind != Kind.AT) continue
		for (keyword in ANCHOR_KEYWORDS) {
			val end = i + 1 + keyword.length
			if (end >= tokens.size || tokens[end].kind != Kind.SEP) continue
			val spelled = tokens.subList(i + 1, end).joinToString("") { if (it.kind == Kind.CHAR) it.value else "" }
			if (spelled == keyword) return Anchor(i, keyword, end + 1)
		}
	}
	return null
}

private sealed class MatcherResult {
	data class Parsed(val matcher: Matcher) : MatcherResult()

	data class Failed(val code: String, val offset: Int) : MatcherResult()
}

private fun parseMatcher(tokens: List<Token>, hashOffset: Int): MatcherResult {
	if (tokens.isEmpty()) return MatcherResult.Failed("empty-fragment", hashOffset)

	val anchor = findAnchor(tokens)
	val rangeAt = tokens.indexOfFirst { it.kind == Kind.RANGE }

	if (anchor != null && (rangeAt == -1 || anchor.at < rangeAt)) {
		val text = tokensToText(tokens.subList(0, anchor.at))
		val anchorText = tokensToText(tokens.subList(anchor.after, tokens.size))
		if (text.isEmpty()) return MatcherResult.Failed("empty-match-text", tokens[0].offset)
		if (anchorText.isEmpty()) return MatcherResult.Failed("empty-anchor", tokens[anchor.at].offset)
		return MatcherResult.Parsed(
			if (anchor.keyword == "before") Matcher.Before(text, anchorText) else Matcher.After(text, anchorText),
		)
	}

	if (rangeAt != -1) {
		val from = tokensToText(tokens.subList(0, rangeAt))
		val to = tokensToText(tokens.subList(rangeAt + 1, tokens.size))
		if (from.isEmpty() || to.isEmpty()) return MatcherResult.Failed("empty-range-bound", tokens[rangeAt].offset)
		return MatcherResult.Parsed(Matcher.Range(from, to))
	}

	return MatcherResult.Parsed(Matcher.Text(tokensToText(tokens)))
}

fun tryParseRef(uri: String): RefParseResult {
	// The angle-bracket pair is markdown's destination wrapper, so it comes off only as a PAIR.
	// Stripping a lone trailing one would silently shorten `#Promise<Response>`.
	val bare = uri.trim()
	val wrapped = bare.length >= 2 && bare.startsWith("<") && bare.endsWith(">")
	val unwrapped = if (wrapped) bare.substring(1, bare.length - 1) else bare
	if (!unwrapped.lowercase().startsWith(REF_SCHEME)) return RefParseResult.NotARef

	// Offsets are shifted past the scheme so they index the ref as WRITTEN, matching the TS side and
	// the shared vectors: an error message quotes the whole ref, so a raw lexer offset points short.
	val tokens = lex(unwrapped.substring(REF_SCHEME.length)).map { it.copy(offset = it.offset + REF_SCHEME.length) }
	val hashAt = tokens.indexOfFirst { it.kind == Kind.HASH }
	val scope = if (hashAt == -1) tokens else tokens.subList(0, hashAt)

	// path := char+. Required, which is what stops `ref://:Foo` from quietly promoting its first
	// segment into the path slot.
	val firstSep = scope.indexOfFirst { it.kind == Kind.SEP }
	val pathTokens = if (firstSep == -1) scope else scope.subList(0, firstSep)
	if (pathTokens.isEmpty()) return RefParseResult.Error("path-required", 0)

	// segment := char*, so an empty one merges. That IS the `::` collapse, stated rather than falling
	// out of a filter as a side effect.
	val segments = ArrayList<String>()
	var cursor = firstSep
	while (cursor != -1) {
		var nextSep = -1
		for (i in cursor + 1 until scope.size) {
			if (scope[i].kind == Kind.SEP) {
				nextSep = i
				break
			}
		}
		val text = tokensToText(scope.subList(cursor + 1, if (nextSep == -1) scope.size else nextSep))
		if (text.isNotEmpty()) segments.add(text)
		cursor = nextSep
	}

	val path = tokensToText(pathTokens)
	if (hashAt == -1) return RefParseResult.Ok(Ref(path, segments, null))

	return when (val matcher = parseMatcher(tokens.subList(hashAt + 1, tokens.size), tokens[hashAt].offset)) {
		is MatcherResult.Failed -> RefParseResult.Error(matcher.code, matcher.offset)
		is MatcherResult.Parsed -> RefParseResult.Ok(Ref(path, segments, matcher.matcher))
	}
}

fun parseRef(uri: String): Ref? = (tryParseRef(uri) as? RefParseResult.Ok)?.ref

private fun escapeCodePoint(unit: String): String =
	unit.toByteArray(Charsets.UTF_8).joinToString("") { "%%%02X".format(it.toInt() and 0xFF) }

private fun encode(raw: String, escapes: Set<Char>): String {
	val out = StringBuilder()
	for (unit in codePoints(raw)) {
		val single = unit.length == 1 && (unit[0] in escapes || unit[0].isWhitespace())
		out.append(if (single) escapeCodePoint(unit) else unit)
	}
	return out.toString()
}

private fun serializeMatcher(matcher: Matcher): String {
	fun text(raw: String) = encode(raw, FRAGMENT_ESCAPES)
	return when (matcher) {
		is Matcher.Text -> "#${text(matcher.text)}"
		is Matcher.Before -> "#${text(matcher.text)}@before:${text(matcher.anchor)}"
		is Matcher.After -> "#${text(matcher.text)}@after:${text(matcher.anchor)}"
		is Matcher.Range -> "#${text(matcher.from)}..${text(matcher.to)}"
	}
}

/**
 * The stable identity of a ref, byte-identical to the MCP's `canonicalKey`.
 *
 * Idempotent by construction: every character the reader treats specially is escaped, so re-reading
 * a key yields the same tokens and therefore the same ref.
 */
fun canonicalKey(ref: Ref): String {
	val scope = (listOf(ref.path) + ref.segments).joinToString(":") { encode(it, SCOPE_ESCAPES) }
	return "$REF_SCHEME$scope${ref.matcher?.let { serializeMatcher(it) } ?: ""}"
}

/** Canonicalize a tapped link destination for manifest lookup, or null if it is not a ref. */
fun canonicalizeRefUri(uri: String): String? = parseRef(uri)?.let { canonicalKey(it) }
